Make sure that OAuth callback errors gets handled

timothyej opened the pull request:
Fixes #288.

Errors during social login is today displayed as json (see #288). This PR fixes so that the error is instead displayed as an error on the login form. It also shows a more specific error message when email was not provided.

**Please see *Discussion* before reviewing.**

### Discussion

After I've made this, I found out that there are views for displaying social login errors (e.g. `facebook_login_failed.jade`), but they doesn't seem to be used.

However, I found it better to display the error on the login form as it allows the user to try again.

Question is, what solution should I go with? Implementing the error views or use the login form as this PR does? If the latter, I should probably show the registration form instead if the error occurred during signup, which is not possible to know at the moment.

### How to verify

1. **Checkout `master`.**
2. Clone the [express-stormpath-sample-project](https://github.com/stormpath/express-stormpath-sample-project).
3. Link to express-stormpath (`npm link` and `npm link express-stormpath`)
4. Configure it with a Facebook Directory.
5. Try to login with Facebook, but in the login popup, choose to not share your email address.
6. Verify that you get an error response in form of json.
7. **Checkout this branch.**
8. Try to login again.
9. Verify that you get the error on the login form instead (*see screenshot*)
10. Also verify that the error message has been rewritten to state that must you share your email.
11. Browse to <http://localhost:3000/callbacks/facebook?&access_token=sdssdsd> with an invalid access token.
12. Verify that you get another error message.
13. Try the same for Google and LinkedIn.

### Screenshots

<img width="637" alt="screen shot 2016-02-08 at 13 36 44" src="https://cloud.githubusercontent.com/assets/721091/12885800/11946968-ce69-11e5-94d6-99a68c40891e.png">

typerandom commented on lib/oauth/error-responder.js:
```diff
@@ -0,0 +1,62 @@
+    });
+
+    if (err.code === 7201) {
+      // Stormpath is unable to create or update the account because the
```
I'd place this comment above `if (err.code === 7201) {`. Since it's related to that line.

typerandom commented on lib/views/linkedin_login_form.jade:
```diff
@@ -15,7 +15,9 @@ script(type='text/javascript').
         }
 
-        serializedQueryString += key + '=' + encodeURIComponent(value);
+        if (key !== 'access_token') {
```
I'd have this for loop just continue, instead of wrapping `serializedQueryString += key + '=' + encodeURIComponent(value);`. Also add a comment to why it is neccessary to explicitly strip this query parameter from the URL.

E.g.

```javascript
// Because the access token isn't bla bla bla...
if (key === 'access_token') {
  continue;
}

serializedQueryString += key + '=' + encodeURIComponent(value);
```

typerandom commented on lib/oauth/error-responder.js:
```diff
@@ -0,0 +1,62 @@
+  function renderLoginFormWithError(err) {
+    var view = req.app.get('stormpathConfig').web.login.view;
+    var nextUri = url.parse(req.query.next || '').path;
+    var formActionUri = config.web.login.uri + (nextUri ? ('?next=' + nextUri) : '');
```
What happens if `login.uri` contains a query string? E.g. if it is `?page=login`. Also, you need to escape `nextUri`.

timothyej replied:
Since this line is just copied from the login controller, I think that we should keep the same behaviour and try to refactor them at once (even better is to use the same code, and not copy). But this is too much work for the scope of this PR.

However, I did add the `encodeURIComponent`.

typerandom commented on lib/oauth/error-responder.js:
```diff
@@ -0,0 +1,62 @@
+  var config = req.app.get('stormpathConfig');
+
+  function renderLoginFormWithError(err) {
+    var view = req.app.get('stormpathConfig').web.login.view;
```
It's nicer for readability to split these lines up.

```javascript
var stormpathConfig = req.app.get('stormpathConfig');
var view = stormpathConfig.web.login.view;
```

typerandom commented on lib/oauth/error-responder.js:
```diff
@@ -0,0 +1,62 @@
+var resolveStateToken = require('./common').resolveStateToken;
+
+/**
+ * oauthErrorResponder takes an error object and respondes
```
Remove `oauthErrorResponder` and capitalize `takes`. Also `respondes` should be `responds`.

typerandom commented on lib/oauth/error-responder.js:
```diff
@@ -0,0 +1,62 @@
+
+var helpers = require('../helpers');
+var forms = require('../forms');
+var resolveStateToken = require('./common').resolveStateToken;
```
For readability, split this into two separate lines.

```javascript
var common =  require('./common');
var resolveStateToken = common.resolveStateToken;
```

timothyej replied:
Decided to just use `common.resolveStateToken()` where it's being used.

typerandom commented on lib/oauth/error-responder.js:
```diff
@@ -0,0 +1,62 @@
+    helpers.render(req, res, view, options);
+  }
+
+  switch (accepts) {
```
This seems overly engineered. The two cases to handle should only be HTML and JSON. So the likelihood of having to add support for some other format seems low. I'd rather just have this as a if, also, removes nesting.

```javascript
if (accepts === 'json') {
  return res.status(err.status || 400).json(err);
}

return renderLoginFormWithError(err);
```

timothyej replied:
I was deciding between these two and was a bit sceptical myself. I've updated it to your suggestion now.

typerandom commented on lib/views/facebook_login_form.jade:
```diff
@@ -16,6 +16,8 @@ script(type='text/javascript').
         var queryStr = window.location.search.replace('?', '');
         // TODO make dynamic
         if (queryStr) {
+          // Remove existing access token from query string.
```
Same as my previous comment. Why do we want to remove this? Please add a comment describing why.

timothyej replied:
I agree with you but don't think it's in the scope of this PR :) But I've added the comment and we can refactor this later (there's even a TODO comment to do so ;)).

typerandom commented on lib/controllers/facebook-login.js:
```diff
@@ -36,7 +38,7 @@ module.exports = function (req, res) {
   if (!req.query.access_token) {
     logger.info('A user attempted to log in via Facebook OAuth without specifying an OAuth token.');
-    return res.status(400).json({ message: 'access_token parameter required.' });
+    return oauth.errorResponder(req, res, { message: 'access_token parameter required.' });
```
To be consistent, maybe always pass an error object to `oauth.errorResponder()`? I.e.

```javascript
return oauth.errorResponder(req, res, new Error('access_token parameter required.'));
```

typerandom commented on lib/controllers/facebook-login.js:
```diff
@@ -49,13 +51,13 @@ module.exports = function (req, res) {
     if (err) {
       logger.info('During a Facebook OAuth login attempt, we were unable to fetch the user\'s social account from Stormpath.');
-      return res.status(err.status || 400).json(err);
+      return oauth.errorResponder(req, res, err);
```
Same as previous comment.

typerandom commented on lib/controllers/facebook-login.js:
```diff
@@ -49,13 +51,13 @@ module.exports = function (req, res) {
       if (err) {
         logger.info('During a Facebook OAuth login attempt, we were unable to fetch the user\'s social account from Stormpath.');
-        return res.status(err.status || 400).json(err);
+        return oauth.errorResponder(req, res, err);
```
Same as previous comment.

typerandom commented on lib/controllers/google-login.js:
```diff
@@ -31,7 +33,7 @@ module.exports = function (req, res) {
   if (!req.query.code) {
     logger.info('A user attempted to log in via Google OAuth without specifying an OAuth token.');
-    return res.status(400).json({ message: 'code parameter required.' });
+    return oauth.errorResponder(req, res, { message: 'code parameter required.' });
```
Same as previous comment. Pass an `Error` object instead.

timothyej commented:
@typerandom Thanks for the review, I've addressed all of your comments in new commits :+1: Please review the code again, and also make sure to test the actual functionality (please note that LinkedIn might not work: #305). Also see my notes under *Discussion*.

robertjd commented:
Thanks @timothyej this looks great!  Yes, I think the error should be shown on the login form because that is the only place we render the social buttons, so it makes sense to place them there.

I added a commit, where I made a few small error rendering fixes. See https://github.com/stormpath/express-stormpath/commit/0ac9328e3f07c0053425745782fe309a1b3df63f
